File: scala/src/main/scala/0309_BestTimeToBuyAndSellStockWithCooldown.scala
object BestTimeToBuyAndSellStockWithCooldown extends App {
  /*
    [1,2,3]

buy => sell, hold
sell => nopos
hold => sell, hold
nopos => buy, nopos

i v                          nopos:0:2
0 1.                 buy:-1:2              nopos
1 2          sell:1      hold:-1:3      ....
2 3.         nopos:1  sell:2:3  hold:-1
                             0
brute force

O(2^n), O(n)


dp: top down
O(n), O(n)

i v                      nopos:0 <-
0 1.         buy:-1. sell:?. hold:?, nopos:0  - base case
1 2          buy:-2  sell:1  hold:-1, nopos:0
2 3.         buy:-3  sell:2  hold:-1  nopos:1
                       nopos:2

buy[n]  = nopos[n-1] - p[n]
sell[n] = buy[n-1].max(hold[n-1]) + p[n]
hold[n] = buy[n-1].max(hold[n-1])
nopos[n] = nopos[n-1].max(sell[n-1])

O(n), O(1)

   */
  def maxProfit(prices: Array[Int]): Int = {

    def go(i: Int, buy: Int, sell: Int, hold: Int, nopos: Int): Int = {
      if (i >= prices.length) sell.max(nopos)
      else
        go(
          i + 1,
          nopos - prices(i),
          buy.max(hold) + prices(i),
          buy.max(hold),
          sell.max(nopos)
        )
    }

    val MinValue = -10000
    go(1, -prices(0), MinValue, MinValue, 0)
  }

  assert(maxProfit(Array(1, 2, 3, 0, 2)) == 3)
  assert(maxProfit(Array(1)) == 0)
}

File: rust/src/bin/0309_best_time_to_buy_and_sell_stock_with_cooldown.rs
fn main() {
    assert_eq!(max_profit(vec![1, 2, 3, 0, 2]), 3);
    assert_eq!(max_profit(vec![1]), 0);
}

pub fn max_profit(prices: Vec<i32>) -> i32 {
    let last = prices.len() - 1;
    let mut buy = -prices[last];
    let mut sell = prices[last];
    let mut nopos = 0;
    let mut holding = 0;

    for i in (0..last).rev() {
        let p = prices[i];

        let next_buy = sell.max(holding) - p;
        let next_sell = nopos + p;
        let next_nopos = buy.max(nopos);
        let next_holding = sell.max(holding);

        buy = next_buy;
        sell = next_sell;
        nopos = next_nopos;
        holding = next_holding;
    }

    buy.max(nopos)
}
